### ISSUE TITLE
Improve stochastic optimizers

### ISSUE THREAD

**Issue by accosmin** [accosmin/nano#108]
Need to improve tuning their hyper-parameters because there is a BIG variance in the resulting train/test error.

Possible solutions:
* limit the decay rate: [1e-4, 1e-1] to make sure some progress is performed
* tune on larger number of epochs to have consistent results

**Comment by accosmin** [comment]
There might be a bug in tuning the stochastic optimizers: the selected configuration does not correspond to the minimum training loss.

**Comment by accosmin** [comment]
Fixed the bug related to tuning.
Many more samples are used now for training, thus hyper-parameter tuning is much more robust.

**Comment by accosmin** [comment]
Fixed all stochastic optimizers except AG-based ones.

**Comment by accosmin** [comment]
Fixed AG-based optimizers as well.
